File: packages/mirascript/src/vm/lib/global/sequence/zip.ts
import { display } from '../../../../helpers/serialize.js';
import { setRecord } from '../../../../helpers/utils.js';
import { Cp } from '../../../checkpoint.js';
import { isVmArray, type VmConst, type VmArray, type VmRecord } from '../../../types/index.js';
import { VmLib, throwError } from '../../helpers.js';
import { entries } from './entries.js';

export const zip = VmLib(
    (data) => {
        const ets = entries(data) as Array<{ 0: string | number; 1: VmArray }>;
        let length = 0;
        for (const { 0: key, 1: arr } of ets) {
            if (!isVmArray(arr)) {
                throwError(`data[${display(key)}] is not an array: ${display(arr)}`, null);
            }
            length = Math.max(length, arr.length);
        }
        if (length === 0) return [];
        const result: Array<VmArray | VmRecord> = [];
        if (isVmArray(data)) {
            for (let i = 0; i < length; i++) {
                Cp();
                const obj: VmConst[] = [];
                for (const { 0: key, 1: arr } of ets) {
                    obj[key as number] = arr[i] ?? null;
                }
                result.push(obj);
            }
        } else {
            for (let i = 0; i < length; i++) {
                Cp();
                const obj: Record<string, VmConst> = {};
                for (const { 0: key, 1: arr } of ets) {
                    setRecord(obj, key as string, arr[i]);
                }
                result.push(obj);
            }
        }
        return result;
    },
    {
        summary: '将数组的数组/记录转换为数组/记录的数组',
        params: { data: { type: 'array | record', description: '要转换的数组/记录' } },
        returns: { type: '(array | record)[]' },
        examples: [
            'zip((x: [1, 2], y: ["a", "b"])) // [(x: 1, y: "a"), (x: 2, y: "b")]',
            `zip([[1, 2], ["a", "b"]]) // [[1, "a"], [2, "b"]]`,
        ],
    },
);

File: crates/python/mirascript/_vm/lib/vm_global/sequence/zip.py
from __future__ import annotations
from typing_extensions import cast

from ....._helpers.serialize import display
from ....._helpers.types import is_vm_array
from ....types.types import VmAny, VmArray, VmRecord
from ....operations.cp import Cp
from ..._helpers import _throw_error
from .entries import entries


def zip(data: VmAny = None) -> VmArray:
    ets = entries(data)
    length = 0
    for el in ets:
        key = el["0"]
        arr = el["1"]
        if not is_vm_array(arr):
            return _throw_error(
                f"data[{display(key)}] is not an array: {display(arr)}", None
            )

        length = max(length, len(arr))

    if length == 0:
        return []

    result: VmArray = []
    is_arr = is_vm_array(data)

    for i in range(length):
        Cp()
        obj: VmArray | VmRecord = [] if is_arr else {}
        for el in ets:
            key = el["0"]
            arr = el["1"]
            val = arr[i] if i < len(arr) else None
            if is_arr:
                cast(VmArray, obj).append(val)
            else:
                cast(VmRecord, obj)[key] = val
        result.append(obj)

    return result
